README
WWW-Desk

Perl Client to make Desk.com API calls.

http://dev.desk.com/API/using-the-api/#general

Desk.com currently support Basic HTTP Authentication and OAuth 1.0a authentication.
WWW::Desk::Auth::oAuth and WWW::Desk::Auth::HTTP will handle authentication


SYNOPSIS

    use WWW::Desk;
    use WWW::Desk::Auth::HTTP;

    my $auth = WWW::Desk::Auth::HTTP->new(
        'username' => 'desk username',
        'password' => 'desk password'
    );
    my $desk = WWW::Desk->new(
        'desk_url'       => 'https://your.desk.com/',
        'authentication' => $auth,
    );

    my $response = $desk->call('/cases','GET', {'locale' => 'en_US'} );

METHOD

    sub call {
        my ( $self, $url_fragment, $http_method, $params ) = @_;
        ...
    }

    Call method accepts
        $url_fragment - API fragment url
        $http_method  - HTTP method, Only supported GET, POST, PATCH, DELETE
        $params       - Additional Parameters which you want to send as query parameters


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc WWW::Desk

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Desk

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/WWW-Desk

    CPAN Ratings
        http://cpanratings.perl.org/d/WWW-Desk

    Search CPAN
        http://search.cpan.org/dist/WWW-Desk/


LICENSE AND COPYRIGHT

Copyright (C) 2014 binary.com

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for WWW-Desk

0.01    30/07/2014
        First version, released on an unsuspecting world.

0.02    31/07/2014
        Added missing test files

0.03    01/08/2014
        Demo application for oAuth work and oAuth class improvement

0.04    01/08/2014
        oAuth test case

0.05    06/08/2014
        more test coverage

0.06    21/08/2014
        CPAN fail test fixes


Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'WWW::Desk',
    AUTHOR             => q{binary.com <rakesh@binary.com>},
    VERSION_FROM       => 'lib/WWW/Desk.pm',
    ABSTRACT_FROM      => 'lib/WWW/Desk.pm',
    LICENSE            => 'Artistic_2_0',
    PL_FILES           => {},
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => { 'ExtUtils::MakeMaker' => 0, },
    BUILD_REQUIRES     => {
	    'Test::NoWarnings'   => 0,
        'Test::More'         => 0,
        'Test::FailWarnings' => 0,
        'Test::Exception'    => 0,
    },
    PREREQ_PM => {
	'IO::Socket::SSL'    => 1.84,
        'Moose'              => 0,
        'Mojo::UserAgent'    => 0,
        'Mojo::Headers'      => 0,
        'Tie::Hash::LRU'     => 0,
        'Mojo::Path'         => 0,
        'Mojo::URL'          => 0,
        'Net::OAuth::Simple' => 0,
        'MIME::Base64'       => 0,
        'Mojo::JSON'         => 0,
        'Carp'               => 0,
    },
    (
        !eval { ExtUtils::MakeMaker->VERSION(6.46) }
        ? ()
        : (
            META_ADD => {
                resources => {
                    homepage   => 'https://github.com/binary-com/perl-WWW-Desk',
                    repository => 'git@github.com:binary-com/perl-WWW-Desk.git',
                    bugtracker =>
                      'http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Desk',
                },
            }
        )
    ),
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES    => 'WWW-Desk-*' },
);


lib/WWW/Desk/Browser.pm
package WWW::Desk::Browser;

use 5.006;
use strict;
use warnings FATAL => 'all';

use Moose;
use Mojo::UserAgent;
use Mojo::URL;
use Mojo::Path;
use Mojo::JSON;

=head1 NAME

WWW::Desk::Browser - Desk.com Browser Client

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 ATTRIBUTES

=head2 base_url

REQUIRED - your desk url

=cut

has 'base_url' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1
);

has 'api_version' => (
    is      => 'ro',
    isa     => 'Str',
    lazy    => 1,
    default => sub {
        return "v2";
    }
);

has 'browser' => (
    is      => 'ro',
    isa     => 'Mojo::UserAgent',
    lazy    => 1,
    default => sub {
        my ($self) = @_;
        my $browser = Mojo::UserAgent->new;
        return $browser;
    }
);

has 'json' => (
    is      => 'ro',
    isa     => 'Mojo::JSON',
    lazy    => 1,
    default => sub {
        my ($self) = @_;
        return Mojo::JSON->new();
    }
);

=head1 SYNOPSIS

    use WWW::Desk::Browser;

    my $browser_client = WWW::Desk::Browser->new();

=head1 SUBROUTINES/METHODS

=head2 prepare_url

Utility method to build base url and path

=cut

sub prepare_url {
    my ( $self, $path ) = @_;
    my $api_version = $self->api_version;
    my $new_path    = Mojo::Path->new($path);
    $path = $new_path->leading_slash(0);
    my $url =
      Mojo::URL->new( $self->base_url )->path("/api/$api_version/$path")
      ->to_abs();
    return $url;
}

=head2 js_encode

Utility method to encode as JSON format

=cut

sub js_encode {
    my ( $self, $response ) = @_;
    return $self->json->encode($response);
}

=head1 AUTHOR

binary.com, C<< <rakesh at binary.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-desk at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Desk>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Desk


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Desk>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Desk>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Desk>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Desk/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 binary.com.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

no Moose;
__PACKAGE__->meta->make_immutable();

1;    # End of WWW::Desk


lib/WWW/Desk/Auth/HTTP.pm
package WWW::Desk::Auth::HTTP;

use 5.006;
use strict;
use warnings FATAL => 'all';

use Moose;
use MIME::Base64;
use Mojo::Headers;

=head1 NAME

WWW::Desk::Auth::HTTP - Desk.com HTTP Basic Authentication

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 ATTRIBUTES

=head2 username

REQUIRED - desk.com username

=cut

has 'username' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1
);

=head2 password

REQUIRED - desk.com password

=cut

has 'password' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1
);

has 'http_header' => (
    is      => 'ro',
    isa     => 'Mojo::Headers',
    lazy    => 1,
    default => sub {
        my ($self) = @_;
        my $http_header = Mojo::Headers->new;
        return $http_header;
    }
);

=head1 SYNOPSIS

    use WWW::Desk::Auth::HTTP;

    my $auth = WWW::Desk::Auth::HTTP->new(
        'username' => 'desk username',
        'password' => 'desk password'
    );
    $auth->login_headers();

=head1 SUBROUTINES/METHODS

=head2 login_headers

Return Mojo::Headers for basic http authentication

=cut

sub login_headers {
    my ($self) = @_;
    my $http = $self->http_header;
    $http->accept('application/json');
    $http->content_type('application/json');
    $http->add( 'Authorization' => 'Basic '
          . encode_base64( $self->username . ':' . $self->password, '' ) );
    return $http;
}

=head1 AUTHOR

binary.com, C<< <rakesh at binary.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-desk at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Desk>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Desk


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Desk>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Desk>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Desk>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Desk/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 binary.com.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

no Moose;
__PACKAGE__->meta->make_immutable();

1;    # End of WWW::Desk::Auth::HTTP


lib/WWW/Desk/Auth/oAuth.pm
package WWW::Desk::Auth::oAuth;

use 5.006;
use strict;
use warnings;

use Moose;
use Mojo::Path;
use Mojo::URL;
use Net::OAuth::Client;
use Tie::Hash::LRU;

=head1 NAME

WWW::Desk::Auth::oAuth - Desk.com oAuth Authentication

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';

our %session;
our $lru = tie %session, 'Tie::Hash::LRU', 100;

=head1 ATTRIBUTES

=head2 api_key

REQUIRED - desk.com api key

=cut

has 'api_key' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1
);

=head2 secret_key

REQUIRED - desk.com api secret key

=cut

has 'secret_key' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1
);

=head2 desk_url

REQUIRED - your desk url

=cut

has 'desk_url' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1,
);

=head2 callback_url

REQUIRED - desk.com oauth callback URI

It must be a URI object

=cut

has 'callback_url' => (
    is       => 'ro',
    isa      => 'URI',
    required => 1,
);

=head2 debug

debug oAuth Requests - boolean type

=cut

has 'debug' => (
    is       => 'ro',
    isa      => 'Bool',
    lazy     => 1,
    default  => sub {
        return 0;
    }
);

=head2 api_version

desk.com api version

=cut

has 'api_version' => (
    is      => 'ro',
    isa     => 'Str',
    lazy    => 1,
    default => sub {
        return "v2";
    }
);

=head2 auth_client

Net::OAuth::Client OAuth protocol object wrapper

=cut

has 'auth_client' => (
    is         => 'ro',
    isa        => 'Net::OAuth::Client',
    lazy_build => 1
);

sub _build_auth_client {
    my ($self) = @_;
    return Net::OAuth::Client->new(
            $self->api_key,
            $self->secret_key,
            protocol_version => '1.0a',
	    site               => $self->desk_url,
            authorize_path     => '/oauth/authorize',
            request_token_path => '/oauth/request_token',
            access_token_path  => '/oauth/access_token',
	    callback           => $self->callback_url,
	    session 	       => \&_session,
            debug              => $self->debug
    );
}

sub _session{
    my ( @data ) = @_;
    if ( $data[0] && $data[1] ) {
        %session = ( $data[0] => $data[1] );
        return %session;
    }
    return $session{$data[0]};
}

=head1 SYNOPSIS

    use WWW::Desk::Auth::oAuth;

    my $auth = WWW::Desk::Auth::oAuth->new(
        'api_key'      => 'api key',
        'secret_key'   => 'secret key',
        'desk_url'     => 'https://my.desk.com',
        'callback_url' => 'https://myapp.com/callback'
    );

    # Visit authorization_url, approve it
    $auth->authorization_url;

    # Use the auth code to fetch the access token
    my $access_token =  $auth->request_access_token(params->{oauth_token}, params->{oauth_verifier});

    # Use the access token to fetch a protected resource
    my $response = $access_token->get( $auth->build_api_url('/customers') );

NOTE: Checkout demo/oAuth_demo.pl for oauth demo application


=head1 SUBROUTINES/METHODS

=head2 authorization_url

Authorization url the user needs to visit to authorize 

=cut

sub authorization_url {
    my ($self) = @_;
    return $self->auth_client->authorize_url;
}

=head2 request_access_token

Request the access token and access token secret for this user.

The user must have authorized this app at the url given by authorization_url first.

Returns the access token and access token secret but also sets them internally so that after calling this method you can immediately call a restricted method.

It accept two parameters $oauth_token, $oauth_verifier.

=cut

sub request_access_token {
    my ($self, $oauth_token, $oauth_verifier) = @_;
    $self->{'oauth_token'} = $oauth_token;
    $self->{'oauth_verifier'} = $oauth_verifier;
    return $self->auth_client->get_access_token($oauth_token, $oauth_verifier);
}

=head2 build_api_url

It build the api abosulte url with the path your supplied

Takes path as input format

=cut

sub build_api_url {
    my ( $self, $path ) = @_;
    my $api_version = $self->api_version;
    my $new_path    = Mojo::Path->new($path);
    $path = $new_path->leading_slash(0);
    my $url =
      Mojo::URL->new( $self->desk_url )->path("/api/$api_version/$path")
      ->to_abs();
    return $url;
}

=head1 AUTHOR

binary.com, C<< <rakesh at binary.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-desk at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Desk>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Desk


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Desk>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Desk>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Desk>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Desk/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 binary.com.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

no Moose;
__PACKAGE__->meta->make_immutable();

1;    # End of WWW::Desk::Auth::oAuth


lib/WWW/Desk.pm
package WWW::Desk;

use 5.006;
use strict;
use warnings FATAL => 'all';

use Moose;
use Carp;
use WWW::Desk::Browser;

=head1 NAME

WWW::Desk - Desk.com perl API

=head1 VERSION

Version 0.06

=cut

our $VERSION = '0.06';

=head1 SYNOPSIS

WWW::Desk will allow you make all API calls using HTTP or oAuth authentication

    use WWW::Desk;
    use WWW::Desk::Auth::HTTP;

    my $auth = WWW::Desk::Auth::HTTP->new(
        'username' => 'desk username',
        'password' => 'desk password'
    );
    my $desk = WWW::Desk->new(
        'desk_url'       => 'https://your.desk.com/',
        'authentication' => $auth,
    );
    my $response = $desk->call('/cases','GET', {'locale' => 'en_US'} );


NOTE: Checkout demo/oAuth_demo.pl for oauth demo application

=cut

=head1 METHOD

    sub call {
        my ( $self, $url_fragment, $http_method, $params ) = @_;
        ...
    }

    Call method accepts
        $url_fragment - API fragment url
        $http_method  - HTTP method, Only supported GET, POST, PATCH, DELETE
        $params       - Additional Parameters which you want to send as query parameters


=head1 ATTRIBUTES

=head2 desk_url

REQUIRED - your desk url

=cut

has 'desk_url' => (
    is       => 'ro',
    isa      => 'Str',
    required => 1
);

has 'browser_client' => (
    is         => 'ro',
    isa        => 'WWW::Desk::Browser',
    lazy_build => 1,
);

sub _build_browser_client {
    my ($self) = @_;
    return WWW::Desk::Browser->new( base_url => $self->desk_url );
}

=head2 authentication

REQUIRED - WWW::Desk::Auth::HTTP or WWW::Desk::Auth::oAuth object

=cut

has 'authentication' => (
    is       => 'ro',
    isa      => 'Object',
    required => 1
);

=head1 SUBROUTINES/METHODS

=head2 call

call method will allow you to make API calls.
url fragment, http method are required to make call

you can also pass params next to http method to add additional paramerters to the url

=cut

sub call {
    my ( $self, $url_fragment, $http_method, $params ) = @_;

    if ( not defined $params ) {
        $params = { 't' => time() };
    }

    return $self->_prepare_response( "400",
        "Argument must be supplied as HASH" )
      unless ref $params eq 'HASH';

    return $self->_prepare_response( "400",
        "Invalid HTTP method. Only supported GET, POST, PATCH, DELETE" )
      unless $http_method =~ /^GET$|^POST$|^PATCH$|^DELETE$/i;

    $http_method = lc $http_method;
    my $browser_client = $self->browser_client;
    my $request_url    = $browser_client->prepare_url($url_fragment);

    my $response;

    my $authentication = $self->authentication;
    if ( ref($authentication) eq 'WWW::Desk::Auth::HTTP' ) {

        my $json_params =
          $params ? $browser_client->js_encode($params) : $params;
        my $http_headers = $self->authentication->login_headers->to_hash();
        $response =
          $browser_client->browser->$http_method(
            $request_url => $http_headers => $json_params );
    }
    elsif ( ref($authentication) eq 'WWW::Desk::Auth::oAuth' ) {
        return $self->_prepare_response( "501",
            "Command line doesn't support oAuth Authentication" );
    }
    else {
        return $self->_prepare_response( "501",
            "Authentication Not Implemented" );
    }

    my $error = $response->error;

    return $self->_prepare_response( 404, $error )
	if ref $error ne 'HASH';

    return $self->_prepare_response( $error->{'code'} || 408,
        $error->{'message'} )
      if $error;

    return $self->_prepare_response( $error->{'code'} || 200,
        $error->{'message'}, $response->res->body );
}

sub _prepare_response {
    my ( $self, $code, $msg, $data ) = @_;
    $data = $self->browser_client->json->decode($data) if $data;
    return {
        'code'    => $code,
        'message' => $msg,
        'data'    => $data
    };
}

=head1 AUTHOR

binary.com, C<< <rakesh at binary.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-desk at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Desk>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Desk


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Desk>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Desk>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Desk>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Desk/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 binary.com.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

no Moose;
__PACKAGE__->meta->make_immutable();

1;    # End of WWW::Desk
